src/Collection.php

```diff
@@ -833,24 +833,21 @@ public function mapReduce(JavascriptInterface $map, JavascriptInterface $reduce,
         }
 
         // Check if the out option is inline because we will want to coerce a primary read preference if not
-        if ( ! is_array($out) && ! is_object($out)) {
-            return false;
-        }
-
-        $out = (array) $out;
-
-        if (key($out) !== 'inline') {
+        if ($this->isOutInline($out)) {
+            $isOutInline = true;
             $options['readPreference'] = new ReadPreference(ReadPreference::RP_PRIMARY);
         }
 
         $server = $this->manager->selectServer($options['readPreference']);
 
-        // Set options if not set already
-        if ( ! isset($options['readConcern']) && ! ($this->readConcern->getLevel() === ReadConcern::MAJORITY) && \MongoDB\server_supports_feature($server, self::$wireVersionForReadConcern)) {
+        /* A "majority" read concern is not compatible with $out inline, so
+         * avoid providing the Collection's read concern if it would conflict.
+         */
+        if ( ! isset($options['readConcern']) && ! ($isOutInline && $this->readConcern->getLevel() === ReadConcern::MAJORITY) && \MongoDB\server_supports_feature($server, self::$wireVersionForReadConcern)) {
```

Comment:
I think there's a logic error in the middle condition:

```
! ($isOutInline && $this->readConcern->getLevel() === ReadConcern::MAJORITY)
````

If we apply [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) to distribute the negation, we get:

```
( ! $isOutInline || $this->readConcern->getLevel() !== ReadConcern::MAJORITY )
```

We can write this in prose as:

> mapReduce is writing to a collection OR read concern is not "majority"

And back in the original condition:

> * the user hasn't specified a read concern AND
> * (mapReduce is writing to a collection OR read concern is not "majority") AND
> * the server supports read concerns

Given this, it's possible that the Collection could have a majority read concern and we'd still inherit it as a default when mapReduce is writing to a collection. Bad news bears.

While we can fix this by negating `$isOutInline` within the middle conditional, it's already hard to read as it is. I'd suggest changing the variable to track the inverse (e.g. `$hasOutputCollection` as mentioned in https://github.com/mongodb/mongo-php-library/pull/419#discussion_r146691277). Then, this condition basically copies the `aggregate()` snippet with one variable rename:

```php
/* A "majority" read concern is not compatible with writing output to a
 * collection, so avoid providing the Collection's read concern if it
 * would conflict.
 */
if ( ! isset($options['readConcern']) &&
     ! ($hasOutputCollection && $this->readConcern->getLevel() === ReadConcern::MAJORITY) &&
    \MongoDB\server_supports_feature($server, self::$wireVersionForReadConcern)) {
    $options['readConcern'] = $this->readConcern;
}
```